### PR TITLE
SearchBar class - add search bars to mods

### PR DESCRIPTION
**Pull request by DeltaNedas**
lets go
not sure if *only* remote servers is good or if i should add it to local and global too

### CODE REVIEW

File: core/src/mindustry/ui/dialogs/JoinDialog.java [hunk 1]
```diff
@@ -22,15 +22,15 @@
...
-    Table hosts = new Table();
-    int totalHosts;
-    int refreshes;
+    public Seq<Server> servers = new Seq<>();
+    public Dialog add;
+    public Server renaming;
+    public Table local = new Table(),
+        remote = new Table(),
+        global = new Table(),
+        hosts = new Table();
```
Comment (summetdev) [inline core/src/mindustry/ui/dialogs/JoinDialog.java:31]: I think it's better to leave old declaration

```suggestion
    public Table local = new Table();
    public Table remote = new Table();
    public Table global = new Table();
    public Table hosts = new Table();
```

File: core/src/mindustry/ui/SearchBar.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import arc.func.*;
+import arc.scene.ui.layout.*;
+import arc.struct.*;
+import mindustry.gen.*;
+
+public class SearchBar{
+    public static <T> Table add(Table parent, Seq<T> list, Func<String, String> queryf,
+            Func<T, String> namef, Cons2<Table, T> itemf){
+        Cons<String>[] rebuild = new Cons[] {null};
+        Table[] pane = new Table[] {null};
```
Comment (summetdev) [inline core/src/mindustry/ui/SearchBar.java:12]: Anuke hate spaces :eyes:

```suggestion
        Cons<String>[] rebuild = new Cons[]{null};
        Table[] pane = new Table[]{null};
```

**Review by summetdev — COMMENTED** [review]
I think it should be used with local and global servers as well. Could you also show a preview? :eyes:

**Comment by DeltaNedas** [comment]
i agree but have no idea how i would do it since theyre h

**Comment by DeltaNedas** [comment]
@Anuken do you agree with summet?

**Comment by DeltaNedas** [comment]
Screenshot_16:28:34
mods works
at this point i should just remove the remote server search bar:
1. messes stuff up
2. still havent implemented my new idea for server search h

**Comment by DeltaNedas** [comment]
@Anuken i have tested now
how would i go about making schematics dialog use SearchBar?

**Comment by joshuaptfan** [comment]
Maybe for mobile screen real estate issues, add a mobile-only button to show the search bar.

**Comment by DeltaNedas** [comment]
better idea: rm -rf android

**Comment by DeltaNedas** [comment]
https://streamable.com/4lmgv1

**Comment by DeltaNedas** [comment]
https://user-images.githubusercontent.com/39013340/102645810-f3ea1b00-415a-11eb-99cf-7ec42956201d.mp4

173kb larger but whatever at least its not quicktime

**Comment by DeltaNedas** [comment]
@Anuken ready

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 3]
```diff
@@ -136,14 +139,12 @@ void setup(){
...
-                table.margin(10f).top();
-
-                boolean anyDisabled = false;
-                for(LoadedMod mod : mods.list()){
-
-                    if(!mod.enabled() && !anyDisabled && mods.list().size > 0){
-                        anyDisabled = true;
+            boolean[] anyDisabled = {false};
+            SearchBar.add(cont, mods.list(),
+                mod -> Strings.stripColors(mod.meta.displayName()),
```
Comment (Anuken) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:144]: The display name already has colors stripped.

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 4]
```diff
@@ -217,20 +218,19 @@ void setup(){
...
-
                     }, Styles.clearPartialt, () -> showMod(mod)).size(w, h).growX().pad(4f);
                     table.row();
-                }
-            });
+                }).margin(10f).top();
 
+                if(mobile && !Core.graphics.isPortrait()){
+                    //hide search bar on mobile, takes up too much space
+                    cont.getCells().get(2).get().clear();
```
Comment (Anuken) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:227]: This is very hacky and will break without an obvious cause if any relevant layout is changed.

**Review by Anuken — CHANGES_REQUESTED** [review]

File: core/src/mindustry/ui/dialogs/ModsDialog.java (hunk 3 above)
Comment (DeltaNedas) [reply core/src/mindustry/ui/dialogs/ModsDialog.java:144]: yes its an old pr

File: core/src/mindustry/ui/dialogs/ModsDialog.java (hunk 4 above)
Comment (DeltaNedas) [reply core/src/mindustry/ui/dialogs/ModsDialog.java:227]: ok ill add a cons for search bar table

**Review by DeltaNedas — COMMENTED** [review]
a

**Comment by DeltaNedas** [comment]
@Anuken

File: core/src/mindustry/ui/dialogs/ModsDialog.java [hunk 5]
```diff
@@ -30,6 +30,9 @@ public ModsDialog(){
         buttons.button("@mods.guide", Icon.link, () -> Core.app.openURI(modGuideURL)).size(210, 64f);
 
         shown(this::setup);
+        if(mobile){
```
Comment (Anuken) [inline core/src/mindustry/ui/dialogs/ModsDialog.java:33]: probably should happen on desktop too

Comment (DeltaNedas) [reply core/src/mindustry/ui/dialogs/ModsDialog.java:33]: but desktop is always landscape

Comment (Anuken) [reply core/src/mindustry/ui/dialogs/ModsDialog.java:33]: right, mistook it for portrait

File: core/src/mindustry/ui/SearchBar.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import arc.func.*;
+import arc.scene.ui.layout.*;
+import arc.struct.*;
+import mindustry.gen.*;
+
+public class SearchBar{
+    public static <T> Table add(Table parent, Seq<T> list, Func<String, String> queryf,
+            Func<T, String> namef, Cons2<Table, T> itemc, Cons<Cell<Table>> barc){
+        Cons<String>[] rebuild = new Cons[]{null};
+        Table[] pane = new Table[]{null};
```
Comment (Anuken) [inline core/src/mindustry/ui/SearchBar.java:12]:
```suggestion
        Cons<String>[] rebuild = {null};
        Table[] pane = {null};
```

**Pull request merged**